docs(TF-409): deployment configuration (#7398)
vuestorefront/vue-storefront: docs/content/guides/6.multistore/2.tooling-and-concepts/4.deployment/7.configuration.md

# Deployment - Configuration

Deploying your Alokai stores is incredibly straightforward! With our powerful deployment system, you can take your store from development to production in minutes. Whether you're managing a single store or orchestrating multiple brands, Alokai's configuration system makes deployment a breeze. This guide will show you how to leverage these capabilities and get your stores up and running with minimal effort.

**What You'll Learn**

::list{type="success"}
- Understanding the role of `alokai.config.json` in deployment

> - Configuring project names and their impact on deployment

Reviewer: Those two from my perspective it's the same only named different ;)

Author: right :) https://github.com/vuestorefront/vue-storefront/pull/7398/commits/89120e4cbf279a312ea24ba4ecca08ccad9cd32f

- Project identification
- Deployment regions
- Framework selection
- Store-specific configurations

Example configuration:
```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "projectName": "fashion-brand-prod",
        "region": "us-east-1",
        "framework": "nextjs"
      }
    },
    "sports-brand": {
      "deployment": {
        "projectName": "sports-brand-prod",
        "region": "eu-west-1",
        "framework": "nuxt"
      }
    }
  }
}
```

::warning Configuration Required
Each deployable store must have a valid configuration in `alokai.config.json`. Template stores shouldn't be placed in the `alokai.config.json` file as they are not deployed.
::

## Configuration Parameters

### `projectName`

The `projectName` parameter is a crucial identifier that links your store to Alokai Console.

Example configuration:
```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "projectName": "fashion-brand-prod",  // Must match Alokai Console project name
        // ... other parameters
      }
    }
  }
}
```

> 

Reviewer: Here, we should put a little explanation how to find it or link to the Console docs with such explanation

Author: Reference added https://github.com/vuestorefront/vue-storefront/pull/7398/commits/100e831f9686c79c719921725f5f56ca3d5f6058

### `region`

The `region` parameter determines where your store will be deployed. 

::warning `region` Required

> Deployment cannot be triggered without a valid region configuration. Ensure you select a region that complies with your data residency requirements.

Reviewer: > Ensure you select a region that complies with your data residency requirements.

Generally good tip, but IMO in bad context. User can't select the region. He needs to match to it, already have in Console. Otherwise, the deployments fail with lack of given instance name in given region

Reviewer: At least that's my impression

Author: ok, I emphasize now, that the region should match the region chosen in the Alokai Console https://github.com/vuestorefront/vue-storefront/pull/7398/commits/1ba9003d1f3414d46cd1676ae00681afdf772014

::

Example region configuration:
```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "region": "us-east-1",  // AWS region identifier
        // ... other parameters
      }
    }
  }
}
```

### `framework`

The `framework` parameter defines whether a store uses Next.js or Nuxt. Alokai supports both frameworks out of the box through two base applications:
- `apps/storefront-unified-nextjs` for Next.js stores
- `apps/storefront-unified-nuxt` for Nuxt stores

Each store in `alokai.config.json` must specify which framework it will use for deployment:

```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "framework": "nextjs",  // This store will deploy using Next.js
        // ... other parameters
      }
    },
    "sports-brand": {
      "deployment": {
        "framework": "nuxt",  // This store will deploy using Nuxt
        // ... other parameters
      }
    }
  }
}
```

> 

Reviewer: Here we should put info box - that explains that this field is not necessary when you have only 1 framework in project. Nothing needs to be defined, CLI automatically matches the framework.

Author: good catch, I forgot about the auto-detection https://github.com/vuestorefront/vue-storefront/pull/7398/commits/26dabf7b990052f2a3cb0c9f1854fc177551af8a

Your project structure can support both frameworks simultaneously. For example:
```bash
apps/
├── storefront-unified-nextjs/     # Base Next.js implementation
├── storefront-unified-nuxt/       # Base Nuxt implementation
└── stores/
    ├── fashion-brand/            
    │   ├── storefront-unified-nextjs/  # Next.js Storefront
    └── sports-brand/
        └── storefront-unified-nuxt/    # Nuxt Storefront
```

This dual-framework support is particularly beneficial when:
- Different teams specialize in different frameworks
- You're migrating from one framework to another
- Specific stores have requirements better suited to a particular framework

## Local Test Deployments

::warning Development Environment Only

> Local deployments should only be used for development and testing purposes. For production deployments, we recommend using GitHub Workflows as they provide better credentials' security, a controlled deployment environment, version control integration, and automated validation checks. You can read more about it in the [CI/CD guide](/guides/multistore/tooling-and-concepts/deployment/ci-cd).

Reviewer: > we recommend using GitHub Workflows

Maybe this term should be more open - like recommending using CI providers such as: GitHub Actions, GitLab CI, Bitbucket CI, Travis CI, Jenkins etc.

Author: definitely https://github.com/vuestorefront/vue-storefront/pull/7398/commits/dd5cdfcee09462208aec7b03e9dbb6df6663d47e

::

One of the great things about Alokai is how simple it makes deployments! The basic deployment command is:

```bash
yarn store deploy --store-id=fashion-brand \
  --cloud-username=your_username \
  --cloud-password=your_password
```

::tip Environment Variables
You can simplify the command by setting environment variables:
```bash
export CLI_CLOUD_USERNAME=your_username
export CLI_CLOUD_PASSWORD=your_password

# Now you only need to specify store-id and registry
yarn store deploy --store-id=fashion-brand
```
::

> 

Reviewer: Might a good place to mention about `.env` file support and that it need to be put on the root-level - the same as `alokai.config.json`

# Deployment - Configuration

> Deploying your Alokai stores is incredibly straightforward! With our powerful deployment system, you can take your store from development to production in minutes. Whether you're managing a single store or orchestrating multiple brands, Alokai's configuration system makes deployment a breeze. This guide will show you how to leverage these capabilities and get your stores up and running with minimal effort.

Reviewer 2: ```suggestion
Deploying your Alokai stores is incredibly straightforward! With our powerful deployment system, you can take your store from development to production in minutes. Whether you're managing a single store or orchestrating multiple brands, Alokai's configuration system makes deployment a breeze. This guide will show you how to leverage these capabilities and get your stores up and running with minimal effort.
```

A bit too much like an add without ay specifics why. I'd mention that briefly and then it'sperfect

Author: any suggestion?

Reviewer: Maybe something like that:

Deploying your Alokai stores is incredibly simple! With our powerful deployment system, you can go from development to production in just a few minutes.

A single CLI command takes care of everything – it builds the store, optimizes it for deployment, creates a Docker image, pushes it to our repository, and initiates the deployment process on the cluster.

Whether you’re managing a single store or orchestrating multiple brands, Alokai’s configuration system makes deployment effortless. This guide will show you how to leverage these capabilities and get your stores up and running with minimal effort.

Reviewer 2: perfect, thanks!

Author: added https://github.com/vuestorefront/vue-storefront/pull/7398/commits/8ee4724362352b04d74cb62afcff201c7494d523

- Store-specific configurations

Example configuration:
```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "projectName": "fashion-brand-prod",
        "region": "us-east-1",
        "framework": "nextjs"
      }
    },
    "sports-brand": {
      "deployment": {
        "projectName": "sports-brand-prod",
        "region": "eu-west-1",
        "framework": "nuxt"
      }
    }
  }
}
```

::warning Configuration Required
Each deployable store must have a valid configuration in `alokai.config.json`. Template stores shouldn't be placed in the `alokai.config.json` file as they are not deployed.
::

## Configuration Parameters

### `projectName`

The `projectName` parameter is a crucial identifier that links your store to Alokai Console.

::tip Finding Your Project Name
You can find your project name in the [Deployment Variables](/console/instance/settings/deployment-variables) section of the Alokai Console.
::

Example configuration:
```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "projectName": "fashion-brand-prod",  // Must match Alokai Console project name
        // ... other parameters
      }
    }
  }
}
```

> ### `region`

Reviewer 3: It's more about the implementation rather the documentation, but this begs the question: "Why do I have to specify the region when I already have the `projectName`?". I guess, it's easier this way, or the `projectName` is not globally unique.

Author: I think that's because we support (at least theoretically) multi regions, but that's a very good question

Reviewer: @mateuszo Yes, we need it - each instance has assigned region. We support multi-region in meaning that you can have instances per different region. One in US, another in Europe, third in Australia. But each of them are separate instances. Confirmed with Cloud Team 👍

- `apps/storefront-unified-nuxt` for Nuxt stores

Each store in `alokai.config.json` may specify which framework it will use for deployment:

```json
{
  "stores": {
    "fashion-brand": {
      "deployment": {
        "framework": "nextjs",  // This store will deploy using Next.js
        // ... other parameters
      }
    },
    "sports-brand": {
      "deployment": {
        "framework": "nuxt",  // This store will deploy using Nuxt
        // ... other parameters
      }
    }
  }
}
```

Your project structure can support both frameworks simultaneously. For example:
```bash
apps/
├── storefront-unified-nextjs/     # Base Next.js implementation
├── storefront-unified-nuxt/       # Base Nuxt implementation
└── stores/
    ├── fashion-brand/            
    │   ├── storefront-unified-nextjs/  # Next.js Storefront
    └── sports-brand/
        └── storefront-unified-nuxt/    # Nuxt Storefront
```

This dual-framework support is particularly beneficial when:
- Different teams specialize in different frameworks
- You're migrating from one framework to another
- Specific stores have requirements better suited to a particular framework

> ## Local Test Deployments

Reviewer 3: The name "Local Test Deployment" is misleading. My first impression was that we deploy it to local docker not the cloud. What it does in fact is: build a storefront locally and deploy from the command line.
Maybe we could call it: "Local build & deployment via CLI"

Author: Updated https://github.com/vuestorefront/vue-storefront/pull/7398/commits/27ded587eb3ffb0c55826066812100243916b669

::warning Development Environment Only
Local deployments should only be used for development and testing purposes. For production deployments, we recommend using CI/CD pipelines (like GitHub Actions, GitLab CI, or Bitbucket Pipelines) as they provide better credentials' security, a controlled deployment environment, version control integration, and automated validation checks. You can read more about it in the [CI/CD guide](/guides/multistore/tooling-and-concepts/deployment/ci-cd).
::

### Prerequisites

1. **Docker**
   - Docker must be installed and running on your machine

2. **Cloud Credentials**
   - Find your credentials in the [Alokai Console](/console/instance/settings/deployment-variables)

>    - Use staging projects for local testing

Reviewer 3: What is a "staging project"?

Author: oh, I was thinking about staging environment xd changed https://github.com/vuestorefront/vue-storefront/pull/7398/commits/feac843d800fec8f2eb12c935ff5fbd6514a83b9

   - Can be customized using `--docker-registry-url` flag or `CLI_DOCKER_REGISTRY_URL` environment variable

4. **Configuration Verification**
   - Double-check `alokai.config.json` configuration
   - Verify `projectName` matches intended deployment target
   - Confirm `region` selection is appropriate for testing

### Deployment Command

One of the great things about Alokai is how simple it makes deployments! The basic deployment command is:

```bash
yarn store deploy --store-id=fashion-brand \
  --cloud-username=your_username \
  --cloud-password=your_password
```

::tip Environment Variables
You can simplify the command by using environment variables. The Alokai CLI automatically loads variables from the `.env` file in your project root, which is the recommended approach:

```bash
# .env file in your project root
CLI_CLOUD_USERNAME=your_username
CLI_CLOUD_PASSWORD=your_password
```

Alternatively, you can set them manually in your terminal:
```bash
export CLI_CLOUD_USERNAME=your_username
export CLI_CLOUD_PASSWORD=your_password
```

With either approach, you'll only need to specify the store-id:
```bash
yarn store deploy --store-id=fashion-brand
```
::

> That's it! The CLI handles all the complexity of building, packaging, and deploying your store. You can also add `--verbose` flag for more control:

Reviewer 3: ```suggestion
That's it! The CLI handles all the complexity of building, packaging, and deploying your store. You can also add `--verbose` flag to see detailed logs:
```
